File: packages/elixir/bowline_client/lib/bowline_client/rules.ex
defmodule BowlineClient.Rules do
  @moduledoc "Validation rules with the same semantics and messages as the Bowline server."

  alias BowlineClient.Issue

  @email ~r/^[^@\s]+@[^@\s]+\.[^@\s]+$/
  @uuid ~r/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/

  @spec required([String.t()], term()) :: [Issue.t()]
  def required(path, value) do
    if empty?(value), do: [Issue.new(path, "required", "is required")], else: []
  end

  @spec min([String.t()], term(), number()) :: [Issue.t()]
  def min(path, value, bound) do
    {size, unit} = size(value)

    if size < bound,
      do: [Issue.new(path, "min", "must be at least #{format(bound)}#{unit}")],
      else: []
  end

  @spec max([String.t()], term(), number()) :: [Issue.t()]
  def max(path, value, bound) do
    {size, unit} = size(value)

    if size > bound,
      do: [Issue.new(path, "max", "must be at most #{format(bound)}#{unit}")],
      else: []
  end

  @spec len([String.t()], term(), number()) :: [Issue.t()]
  def len(path, value, bound) do
    {size, unit} = size(value)

    if size != bound,
      do: [Issue.new(path, "len", "must be exactly #{format(bound)}#{unit}")],
      else: []
  end

  @spec one_of([String.t()], term(), [String.t()]) :: [Issue.t()]
  def one_of(path, value, options) do
    if to_string(value) in options,
      do: [],
      else: [Issue.new(path, "oneof", "must be one of " <> Enum.join(options, " "))]
  end

  @spec email([String.t()], String.t()) :: [Issue.t()]
  def email(path, value) do
    if email?(value), do: [], else: [Issue.new(path, "email", "must be a valid email address")]
  end

  @spec url([String.t()], String.t()) :: [Issue.t()]
  def url(path, value) do
    if url?(value), do: [], else: [Issue.new(path, "url", "must be a valid URL")]
  end

  @spec uuid([String.t()], String.t()) :: [Issue.t()]
  def uuid(path, value) do
    if uuid?(value), do: [], else: [Issue.new(path, "uuid", "must be a valid UUID")]
  end

  @spec email?(term()) :: boolean()
  def email?(value) when is_binary(value), do: Regex.match?(@email, value)
  def email?(_), do: false

  @spec url?(term()) :: boolean()
  def url?(value) when is_binary(value) do
    case URI.new(value) do
      {:ok, %URI{scheme: scheme, host: host}} when is_binary(scheme) and is_binary(host) ->
        scheme != "" and host != ""

      _ ->
        false
    end
  end

  def url?(_), do: false

  @spec uuid?(term()) :: boolean()
  def uuid?(value) when is_binary(value), do: Regex.match?(@uuid, value)
  def uuid?(_), do: false

  defp empty?(nil), do: true
  defp empty?(""), do: true
  defp empty?(0), do: true
  defp empty?(false), do: true
  defp empty?([]), do: true
  defp empty?(value) when is_float(value), do: value == 0.0
  defp empty?(map) when is_map(map) and not is_struct(map), do: map_size(map) == 0
  defp empty?(_), do: false

  defp size(value) when is_binary(value), do: {String.length(value), " characters"}
  defp size(value) when is_list(value), do: {length(value), " items"}
  defp size(value) when is_map(value) and not is_struct(value), do: {map_size(value), " items"}
  defp size(value) when is_number(value), do: {value, ""}
  defp size(_), do: {0, ""}

  defp format(bound) when is_float(bound) and bound == trunc(bound),
    do: Integer.to_string(trunc(bound))

  defp format(bound), do: to_string(bound)
end

File: packages/rust/bowline-client/src/rules.rs
use crate::{Issue, Validate};
use std::collections::BTreeMap;

fn at(path: &[String], field: &str) -> Vec<String> {
    let mut out = path.to_vec();
    out.push(field.to_string());
    out
}

fn push(path: &[String], field: &str, rule: &str, message: String, issues: &mut Vec<Issue>) {
    issues.push(Issue::new(at(path, field), rule, message));
}

pub fn required_str(v: &str, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if v.is_empty() {
        push(path, field, "required", "is required".into(), issues);
    }
}

pub fn required_int(v: i64, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if v == 0 {
        push(path, field, "required", "is required".into(), issues);
    }
}

pub fn required_float(v: f64, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if v == 0.0 {
        push(path, field, "required", "is required".into(), issues);
    }
}

pub fn required_bool(v: bool, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if !v {
        push(path, field, "required", "is required".into(), issues);
    }
}

pub fn required_len(len: usize, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if len == 0 {
        push(path, field, "required", "is required".into(), issues);
    }
}

pub fn required_some<T>(v: &Option<T>, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if v.is_none() {
        push(path, field, "required", "is required".into(), issues);
    }
}

pub fn min_len(
    len: usize,
    min: usize,
    unit: &str,
    path: &[String],
    field: &str,
    issues: &mut Vec<Issue>,
) {
    if len < min {
        push(
            path,
            field,
            "min",
            format!("must be at least {min}{unit}"),
            issues,
        );
    }
}

pub fn max_len(
    len: usize,
    max: usize,
    unit: &str,
    path: &[String],
    field: &str,
    issues: &mut Vec<Issue>,
) {
    if len > max {
        push(
            path,
            field,
            "max",
            format!("must be at most {max}{unit}"),
            issues,
        );
    }
}

pub fn exact_len(
    len: usize,
    want: usize,
    unit: &str,
    path: &[String],
    field: &str,
    issues: &mut Vec<Issue>,
) {
    if len != want {
        push(
            path,
            field,
            "len",
            format!("must be exactly {want}{unit}"),
            issues,
        );
    }
}

pub fn min_num(
    v: f64,
    min: f64,
    param: &str,
    path: &[String],
    field: &str,
    issues: &mut Vec<Issue>,
) {
    if v < min {
        push(
            path,
            field,
            "min",
            format!("must be at least {param}"),
            issues,
        );
    }
}

pub fn max_num(
    v: f64,
    max: f64,
    param: &str,
    path: &[String],
    field: &str,
    issues: &mut Vec<Issue>,
) {
    if v > max {
        push(
            path,
            field,
            "max",
            format!("must be at most {param}"),
            issues,
        );
    }
}

pub fn exact_num(
    v: f64,
    want: f64,
    param: &str,
    path: &[String],
    field: &str,
    issues: &mut Vec<Issue>,
) {
    if v != want {
        push(
            path,
            field,
            "len",
            format!("must be exactly {param}"),
            issues,
        );
    }
}

pub fn one_of(v: &str, options: &[&str], path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if !options.contains(&v) {
        push(
            path,
            field,
            "oneof",
            format!("must be one of {}", options.join(" ")),
            issues,
        );
    }
}

pub fn one_of_int(v: i64, options: &[&str], path: &[String], field: &str, issues: &mut Vec<Issue>) {
    one_of(&v.to_string(), options, path, field, issues);
}

pub fn chars(v: &str) -> usize {
    v.chars().count()
}

pub fn is_email(v: &str) -> bool {
    let Some((local, domain)) = v.split_once('@') else {
        return false;
    };
    if local.is_empty() || domain.is_empty() || domain.contains('@') {
        return false;
    }
    !v.chars()
        .any(|c| c.is_whitespace() || c == '<' || c == '>' || c == '"' || c == ',')
}

pub fn is_url(v: &str) -> bool {
    let Some((scheme, rest)) = v.split_once("://") else {
        return false;
    };
    let scheme_ok = !scheme.is_empty()
        && scheme
            .chars()
            .next()
            .is_some_and(|c| c.is_ascii_alphabetic())
        && scheme
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '+' || c == '-' || c == '.');
    let host = rest.split(['/', '?', '#']).next().unwrap_or("");
    scheme_ok && !host.is_empty()
}

pub fn is_uuid(v: &str) -> bool {
    let groups: Vec<&str> = v.split('-').collect();
    let lengths = [8, 4, 4, 4, 12];
    groups.len() == 5
        && groups
            .iter()
            .zip(lengths.iter())
            .all(|(g, n)| g.len() == *n && g.chars().all(|c| c.is_ascii_hexdigit()))
}

pub fn email(v: &str, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if !is_email(v) {
        push(
            path,
            field,
            "email",
            "must be a valid email address".into(),
            issues,
        );
    }
}

pub fn url(v: &str, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if !is_url(v) {
        push(path, field, "url", "must be a valid URL".into(), issues);
    }
}

pub fn uuid(v: &str, path: &[String], field: &str, issues: &mut Vec<Issue>) {
    if !is_uuid(v) {
        push(path, field, "uuid", "must be a valid UUID".into(), issues);
    }
}

pub fn nested<T: Validate + ?Sized>(
    v: &T,
    path: &mut Vec<String>,
    field: &str,
    issues: &mut Vec<Issue>,
) {
    path.push(field.to_string());
    v.validate(path, issues);
    path.pop();
}

pub fn each<T: Validate>(
    items: &[T],
    path: &mut Vec<String>,
    field: &str,
    issues: &mut Vec<Issue>,
) {
    path.push(field.to_string());
    for (i, item) in items.iter().enumerate() {
        path.push(i.to_string());
        item.validate(path, issues);
        path.pop();
    }
    path.pop();
}

pub fn each_map<K: ToString, V: Validate>(
    items: &BTreeMap<K, V>,
    path: &mut Vec<String>,
    field: &str,
    issues: &mut Vec<Issue>,
) {
    path.push(field.to_string());
    for (key, item) in items {
        path.push(key.to_string());
        item.validate(path, issues);
        path.pop();
    }
    path.pop();
}
